packages/@aws-cdk/cdk/lib/cloudformation/fn.ts

```diff
@@ -92,60 +98,67 @@ export class FnJoin extends Fn {
     if (listOfValues.length === 0) {
       throw new Error(`FnJoin requires at least one value to be provided`);
     }
-    super('Fn::Join', [ delimiter, listOfValues ]);
+    // Passing the values as a token, optimization requires resolving stringified tokens, we should be deferred until
+    // this token is itself being resolved.
+    super('Fn::Join', [ delimiter, new Token(() => this.resolveValues()) ]);
+    this.delimiter = delimiter;
+    this.listOfValues = listOfValues;
   }
-}
 
-/**
- * Alias for ``FnJoin('', listOfValues)``.
- */
-export class FnConcat extends FnJoin {
-  private readonly listOfValues: any[];
+  public resolve(): any {
+    if (this.resolveValues().length === 1) {
+      return this.resolveValues()[0];
+    }
+    return super.resolve();
+  }
 
   /**
-   * Creates an ``Fn::Join`` function with an empty delimiter.
-   * @param listOfValues The list of values to concatenate.
+   * Optimization: if an Fn::Join is nested in another one and they share the same delimiter, then flatten it up. Also,
+   * if two concatenated elements are literal strings (not tokens), then pre-concatenate them with the delimiter, to
+   * generate shorter output.
    */
-  constructor(...listOfValues: any[]) {
-    // Optimization: if any of the input arguments is also a FnConcat,
-    // splice their list of values into the current FnConcat. 'instanceof'
-    // can fail, but we do not depend depend on this for correctness.
-    //
-    // Do the same for resolved intrinsics, so we can detect this
-    // happening both at Token as well as at CloudFormation level.
+  private resolveValues() {
+    if (this._resolvedValues) { return this._resolvedValues; }
 
+    this._resolvedValues = [...this.listOfValues.map(e => resolve(e))];
```

Comment:
Readability: since this is not the final value of `this._resolvedValues`, I would recommend assigning this to a local variable, mutate it and then at the end of this method assign it to `this._resolvedValues`. It took me 40 seconds to realize this is not the resolved values, but only a beginning of a beautiful journey.

Reply:
Yeah I though of this but somehow it seemed like a good idea to save one assignment. ¯\_(シ)_/¯

Reply:
Yeah I though of this but somehow it seemed like a good idea to save one assignment. ¯\\\_(シ)\_/¯

Reply:
Yeah makes sense

Reply:
Yap that makes total sense!

Reply:
Yah that makes sense

Reply:
Makes sense :)